src/libraries/System.Linq/src/System/Linq/Max.cs

```diff
@@ -15,6 +15,11 @@ public static int Max(this IEnumerable<int> source)
                 ThrowHelper.ThrowArgumentNullException(ExceptionArgument.source);
             }
 
+            if (source.GetType() == typeof(int[]))
```

Comment:
Why not `source is int[] array`?

Reply:
It's measurably more expensive here. This check is as close to free as possible, a simple compare against the type token, whereas the is check ends up incurring a non-inlined call to a helper function, as technically the is can be satisfied by other types, e.g. uint[]. It adds enough overhead that it was showing up at a 10-20% regression on smaller non-inputs. The overhead with the explicit check is small enough that we might consider adding array handling in other such sink methods, even if not vectorizable.